backend/src/controller/chatController.ts

```diff
@@ -43,28 +43,30 @@ function combineChatDefenceReports(
 
 function createNewUserMessages(
 	message: string,
-	transformedMessage: string | null
+	messageTransformation: MessageTransformation | null
 ): ChatHistoryMessage[] {
-	if (transformedMessage) {
-		// if message has been transformed
+	if (messageTransformation) {
 		return [
-			// original message
 			{
 				completion: null,
 				chatMessageType: CHAT_MESSAGE_TYPE.USER,
 				infoMessage: message,
 			},
-			// transformed message
+			{
+				completion: null,
+				chatMessageType: CHAT_MESSAGE_TYPE.INFO,
+				infoMessage: messageTransformation.transformedMessageInfo,
+			},
 			{
 				completion: {
 					role: 'user',
-					content: transformedMessage,
+					content: messageTransformation.transformedMessageCombined,
 				},
 				chatMessageType: CHAT_MESSAGE_TYPE.USER_TRANSFORMED,
+				transformedMessage: messageTransformation.transformedMessage,
```

Comment:
I'm a little unsure about the types now, they feel a little loose. Here we only expect transformedMessage if the type is USER_TRANSFORMED, right? Additionally, completion.role will always be 'user' in that case.

Seems maybe we could split our types for a touch more type-safety, and make ChatHistoryMessage a union type of these.

```
type ChatUserMessageTransformed = {
	completion: ChatCompletionUserMessageParam;
	chatMessageType: CHAT_MESSAGE_TYPE.USER_TRANSFORMED;
	transformedMessage: TransformedChatMessage;
}

type ChatGenericMessage = {
  completion: ChatCompletionMessageParam | null;
  chatMessageType: CHAT_MESSAGE_TYPE;
  infoMessage?: string | null;
}

type ChatMessage = ChatUserMessageTransformed | ChatGenericMessage;
```

In fact, we could go even further and be more specific about _all_ our chat message types in the backend:
- ERROR_MSG and BOT_BLOCKED always have infoMessage but never completion
- SYSTEM always has completion (of type ChatCompletionSystemMessageParam) but never infoMessage
- BOT always has completion (of type ChatCompletionAssistantMessageParam) but never infoMessage. Note our current types are not quite up-to-date; interface "ChatCompletionMessage" which we are currently using in ChatGptReply is actually identical to "ChatCompletionAssistantMessageParam" so we should use the latter to match our other completion types)
- FUNCTION_CALL always has completion (of type ChatCompletionAssistantMessageParam or ChatCompletionToolMessageParam) but never infoMessage
- USER _either_ has infoMessage _or_ has completion (of type ChatCompletionUserMessageParam)

Note there are chat messages that come from the frontend, but they are only a subset of types (DEFENCE_ALERTED, DEFENCE_TRIGGERED, INFO, LEVEL_INFO, RESET_LEVEL), none have completion and all have infoMessage. So we might want another type to represent those front-end chat messages.

Reply:
agreed on teams that it would be a separate PR for the same ticket.

Reply:
Okay, now we have a problem where we need to check that a certain property in a chatMessage exists.

Here's a mini example:
Suppose we have two chat subtypes and a chat type that is the union of the former two.
```typescript
type ChatInfo = {
	info: string;
};

type ChatMessage = {
	message: string;
};

type Chat = ChatInfo | ChatMessage;
```

Now suppose we want to print the message from a given variable of type `chat`. We cannot do simply
```typescript
function printMessage(chat: Chat) {
  console.log(chat.message)
}
```

because `chatInfo` doesn't have the `message` property. Specifically, here's the ts error:

```
Property 'message' does not exist on type 'chat'.
  Property 'message' does not exist on type 'chatInfo'.ts(2339)
```

So instead we have to do something like.
```typescript
function printMessage(chat: Chat) {
  if ('message' in chat) {
    console.log(chat.message)
  }
}
```

Reply:
Changes implemented here #803